themes/default/files/scripts/analysis-display.js

```diff
@@ -0,0 +1,629 @@
+
+            /** The decryption key for attempt data files, loaded from the exam definition.
+             */
+            this.default_passcode = ko.observable(this.exam_object?.navigation?.downloadEncryptionKey);
+
+            /** Should the `overridden_passcode` be used instead of `default_passcode`?
+             */
+            this.override_passcode = ko.observable(false);
+
+            /** The current tab the user is on. Toggles between 'upload', 'list_files', and 'table'.*/
+            this.current_tab = ko.observable('list_files');
+
+            /** The potential options for table display, along with description*/
+            /** Descriptive names for each column in the full table.
+             *  This is an array of four rows, containing cells that span several rows or columns.
+             */
+            this.table_header_readable = ko.observableArray([[]]);
+
+            /** Machine-readable names for each column in the full table, giving the path to the corresponding question/part.
+             */
+            this.table_header_computer = ko.observableArray([]);
+
+            this.table_format_options = ko.observableArray([
+                { label: 'total', name: R('analysis.table.total')},
+                { label: 'question', name: R('analysis.table.question')},
+                { label: 'all', name: R('analysis.table.all')}
+            ]);
+
+            /** The currently selected table format, from the above labels.
+             */
+            this.table_format = ko.observable(this.table_format_options()[0]);
+            
+            /** The uploaded files, sorted by status and then by student name.
+             */
+            this.sorted_files = ko.computed(function() {
+                return this.uploaded_files().slice().sort((a,b) => {
+                    const a_status = a.status();
+                    const b_status = b.status();
+                    if(a_status != b_status) {
+                        return a_status < b_status ? -1 : a_status > b_status ? 1 : 0;
+                    }
+                    function canonical_name(f) {
+                        return f.student_name().toLowerCase().trim();
+                    }
+                    const a_name = canonical_name(a);
+                    const b_name = canonical_name(b);
+
+                    return a_name < b_name ? -1 : a_name > b_name ? 1 : 0;
+                })
+            },this);
+
+            /** Just uploaded files which have been succesfully decrypted. Only these are shown in the results table.
+             */
+            this.decrypted_files = ko.computed(function() {
+                return this.uploaded_files().filter(f => f.status() == 'processed');
+            },this);
+
+            /** The "expected results" row for the "all details" table: available marks for the exam and each question and part.
+             */
+            this.full_expected_results_row = ko.computed(() => {
```

Comment:
I've noticed that if the students have consistent max marks (although the question is not set up to mandate that), the 'all data' display will show the correct total, but the 'question totals' display will show 'varies'. Is it worth having this match?

Reply:
Escalation! If one student has a higher max mark, it takes that rather than using 'varies'.

Reply:
Sorry, I hadn't seen your comment! I think that's still coming from my initial function, see line 544, it sets the max marks but if they are able to be different between attempts it sets it to 'varies'.

Reply:
Exam and Question totals are now showing '0' for expected results! (But All details works!)